java-dynamic-sqs-listener-api/src/main/java/com/jashmore/sqs/retriever/MessageRetriever.java

```diff
@@ -0,0 +1,61 @@
+package com.jashmore.sqs.retriever;
+
+import com.amazonaws.services.sqs.model.Message;
+import com.jashmore.sqs.broker.MessageBroker;
+import com.jashmore.sqs.processor.MessageProcessor;
+
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+
+/**
+ * Class used for retrieving messages to execute from the queue.
+ *
+ * <p>If you were to consider this library as similar to a pub-sub system, this could be considered the publisher.  It polls for messages from the
+ * remote queue which will be taken by the {@link MessageBroker} and transferred to the corresponding {@link MessageProcessor} that knows how to process
+ * this message.
+ *
+ * <p>As there could be multiple threads wanting to process messages the implementations of this class must be thread safe.
+ */
+@ThreadSafe
+public interface MessageRetriever {
+    /**
+     * Retrieve a message from the queue now if there are any available.
+     *
+     * <p>This will not keep waiting until a message is placed onto the queue.
+     *
+     * @return the optional message obtained from the queue
+     * @throws InterruptedException if the thread was interrupted while waiting for a message
```

Comment:
Do you need to throw IEX if its non-blocking call?

Reply:
Hmmm yeah this is a good point. The IndividualMessageRetriever however does do a call out to the service so it is actually blocking for that implementation with IO?

Reply:
It probably makes sense to keep IEX here as that will provide uniform handling for the broker. Regardless of what happens in the retriever, if an interrupt is thrown the broker can handle accordingly.